### PR TITLE
Add better portable PDB caching to System.Diagnostics.StackTrace. 

### PR DESCRIPTION
**Pull request by mikem8361**
Make the pdb reader caching for StackTrace thread safe.

### CODE REVIEW

File: src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs [hunk 1]
```diff
@@ -29,7 +29,10 @@ void IDisposable.Dispose()
         {
             foreach (MetadataReaderProvider provider in _metadataCache.Values)
             {
-                provider.Dispose();
+                if (provider != null)
+                {
+                    provider.Dispose();
+                }
```
Comment (stephentoub) [inline src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: Nit: could just be:
```C#
provider?.Dispose();
```

File: src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs [hunk 2]
```diff
@@ -123,20 +126,25 @@ private unsafe MetadataReader TryGetReader(string assemblyPath, IntPtr loadedPeA
             MetadataReaderProvider provider;
             if (_metadataCache.TryGetValue(cacheKey, out provider))
             {
+                if (provider == null)
+                {
+                    return null;
+                }
```
Comment (stephentoub) [inline src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: Nit: could just be:
```C#
return provider?.GetMetadataReader();
```

File: src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs [hunk 3]
```diff
@@ -123,20 +126,25 @@ private unsafe MetadataReader TryGetReader(string assemblyPath, IntPtr loadedPeA
...
                 return provider.GetMetadataReader();
             }
 
             provider = (inMemoryPdbAddress != IntPtr.Zero) ?
                 TryOpenReaderForInMemoryPdb(inMemoryPdbAddress, inMemoryPdbSize) :
                 TryOpenReaderFromAssemblyFile(assemblyPath, loadedPeAddress, loadedPeSize);
 
+            // This may fail as another thread might have beaten us to it, but it doesn't matter
+            _metadataCache.TryAdd(cacheKey, provider);
```
Comment (stephentoub) [inline src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: We want to add it even if provider is null?

Comment (noahfalk) [reply src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs]: Yep : )
Null means no symbols were found for a module so its a common case in this code path and very valuable performance-wise to cache.

**Comment by mikem8361** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by mikem8361** [comment]
@dotnet-bot test Linux x64 Release Build

**Pull request merged**